scripts/backend.py

```diff
@@ -46,25 +93,151 @@ def random_file_from(dname, cont=None):
+            if 'stream_title' in dj_info and len(dj_info['stream_title']) > 0:
+                status_metadata['album'] = dj_info['stream_title']
+
+    # get the current metadata from icecast
+    sources = req_info['icestats']['source']
+    for source in sources:
+        if LIVESTREAM_INFO['CHANNEL'] in source['listenurl']:
+            # if no server_type then fallback is active, if we're streaming the server_type starts with audio
+            # a regular mpd source's type is application/ogg
+            if 'server_type' in source and 'audio' in source['server_type']:
+                for k in status_metadata:
+                    if k in source:
+                        status_metadata[k] = source[k].strip()
+                break
+
+    # if no last track or it's not the same as the last track append
+    if len(LIVESTREAM_INFO['last_played']) == 0 or LIVESTREAM_INFO['last_played'][-1].title != status_metadata['title']:
+        newest_track = LivestreamTrack(status_metadata['artist'], status_metadata['title'], time.time() + LIVESTREAM_INFO['STREAM_DELAY'])
+        LIVESTREAM_INFO['last_played'].append(newest_track)
+        # we only want 5 last tracks + current
+        if len(LIVESTREAM_INFO['last_played']) > 6:
+            LIVESTREAM_INFO['last_played'].pop(0)
+
+    time_now = time.time()
+
+    pinfo = {
+        'current': status_metadata,
+        'elapsed': int(time_now - LIVESTREAM_INFO['last_played'][-1].first_seen),
+        'before': [],
+        'stream_data': stream_data
+    }
+
+    # populate before, radio expects tracks in most recent to oldest order
+    # and their times need to be calculated because we don't know exact tracklengths
+    indices = range(len(LIVESTREAM_INFO['last_played']) - 2, -1, -1)
+    for i in indices:
+        track = LIVESTREAM_INFO['last_played'][i]
+        track_after = LIVESTREAM_INFO['last_played'][i + 1]
+        prev_metadata = {
+            'artist': track.artist,
+            'title': track.title,
+            'time': (track_after.first_seen - track.first_seen)
+        }
+        pinfo['before'].append(prev_metadata)
+
+    channels[LIVESTREAM_INFO['CHANNEL']]["cache"] = (pinfo, 200)
+
+
+def playlist_info_update_task():
+    global channels, playlist_update_counter
+    for channel in channels:
+        if LIVESTREAM_INFO['active'] and channel == LIVESTREAM_INFO['CHANNEL']:
+            playlist_update_counter = (playlist_update_counter + 1) % 5
+            if playlist_update_counter == 1:
+                update_livestream_info()
+            else:
+                continue
+        else:
+            update_mpd_info(channel)
+
+
+bg_scheduler.add_job(
+    func=playlist_info_update_task,
+    trigger=IntervalTrigger(seconds=1),
+    id='playlist_update',
+    name='Update [channel].json\'s',
+    replace_existing=True)
+
+
+def playlist_for(channel):
+    """Return the playlist of the given channel, as JSON."""
+    global channels
+
+    pinfo = channels[channel]["cache"]
```

Comment:
There's a race condition here, where someone requests the playlist before `update_mpd_info` has happened (even though it's only a second).  You could use `channels[channel].get("cache")`, which defaults to `None` if there is no value.